Implement SETVER.EXE command

FeralChild64 opened the pull request:
This PR implements a `SETVER.EXE` utility, with nearly all extensions found in DR-DOS (this is the only extended version known to me; Microsoft has ditched this tool in modern Windows, so no more extensions should be expected from them).

The `ver set VERSION` command is "deprecated", but still mentioned in the documentation, I do not plan to remove it, ever. Unfortunately, this `ver` extended syntax isn't similar to other DOS commands at all, and it goes against the overall DOS design - but I guess we have to live with it - for sake of compatibility.

Technical reading regarding DOS version reporting and SETVER utility:
[https://www.geoffchappell.com/notes/dos/interrupts/21h/30h/index.htm](https://www.geoffchappell.com/notes/dos/interrupts/21h/30h/index.htm)

dreamer commented on src/dosbox.cpp:
```diff
@@ -1169,6 +1169,12 @@ void DOSBOX_Init()
 	pstring->Set_help(
 	        "Language code of the keyboard layout, or 'auto' ('auto' by default).");
 
+	pstring = secprop->Add_path("setver_table_file", only_at_start, "");
```
Maybe we should just pick a name for the database file, without giving users option for changing the path?

FeralChild64 replied:
I would prefer to have such option, as I am storing the database inside my virtual drive C: - this makes archiving my guest system slightly easier, and I can check version override using DOS Navigator.

dreamer commented on src/dos/program_setver.cpp:
```diff
@@ -0,0 +1,721 @@
+	}
+}
+
+bool SETVER::ParseVersion(const std::string& version_str, FakeVersion& version)
```
I strongly suggest rewriting this function using https://en.cppreference.com/w/cpp/regex

FeralChild64 replied:
Done. The last time I tried to use regular expressions in DOSBox Staging reviewer asked me to migrate away from them. We'll see...

dreamer commented on src/dos/program_setver.cpp:
```diff
@@ -0,0 +1,721 @@
+	// Write table into the file
+	std::ofstream file(file_path.c_str(), std::ios::trunc);
+	if (setver_table.is_global_version_set) {
+		file << "\t" << setver_table.version_global << std::endl;
```
Ugh, using `std::endl` will make this file format non-portable; use Unix line endings for all OSes, please.

FeralChild64 replied:
Done.

kcgen commented on src/dos/dos.cpp:
```diff
@@ -700,29 +708,32 @@ static Bitu DOS_21Handler(void) {
-	case 0x36:		/* Get Free Disk Space */
-		{
-			uint16_t bytes,clusters,free;
+		        case 6: // Get true version number, not affected by SETVER
+		        {
+			        DOS_PSP psp(dos.psp());
+			        reg_bl = dos.version.major;
+			        reg_bh = dos.version.minor;
+			        reg_dl = dos.version.revision;
+			        reg_dh = 0x10; // DOS in HMA
+		        } break;
+		        default:
+			        LOG(LOG_DOSMISC, LOG_ERROR)
+			        ("Weird 0x33 call %2X", reg_al);
+			        reg_al = 0xff;
+			        break;
+		        }
+		        break;
+	case 0x34: /* Get INDos Flag */
+		        SegSet16(es, DOS_SDA_SEG);
+		        reg_bx = DOS_SDA_OFS + 0x01;
+		        break;
+	case 0x35: /* Get interrupt vector */
+		        reg_bx = real_readw(0, ((uint16_t)reg_al) * 4);
+		        SegSet16(es, real_readw(0, ((uint16_t)reg_al) * 4 + 2));
```
I'm having trouble seeing context here - is the indenting on line 725 and prior due to an inner switch statement inside of an outer case?

(Just mentioning it - sometimes the clang formatter can incorrectly format switches)

FeralChild64 replied:
I have reformatted the code manually, in a separate commit for better bisectability.

kcgen commented on src/dos/dos.cpp:
```diff
@@ -700,29 +708,32 @@ static Bitu DOS_21Handler(void) {
+		        break;
+	case 0x36: /* Get Free Disk Space */
+	{
+		        uint16_t bytes,clusters,free;
 			uint8_t sectors;
 			if (DOS_GetFreeDiskSpace(reg_dl,&bytes,&sectors,&clusters,&free)) {
```
(clang formatter might be off here)

FeralChild64 replied:
Reformatted manually, as above.

kcgen commented on src/dos/program_setver.cpp:
```diff
@@ -0,0 +1,744 @@
+	SETVER::FakeVersion version_global = {};
+	bool is_global_version_set         = false;
+
+	NameVersionTable by_file_name = {
+	        // Since MS-DOS 6.22 has some default version table,
+	        // we can provide some sane defaults too
+	        {"WIN100.BIN", {3, 40}}, // fixes Microsoft Windows 1.x
+	        {"WIN200.BIN", {3, 40}}, // fixes Microsoft Windows 2.x
+	};
```
excellent!

kcgen commented on src/dos/program_setver.cpp:
```diff
@@ -0,0 +1,744 @@
+	return true;
+}
+
+bool SETVER::PreprocessName(std::string& name, const bool permissive)
+{
+	// Preprocess file names and relative paths
+	trim(name);
+	if (IsNameWithPath(name)) {
+		char buffer[DOS_PATHLENGTH];
+		if (!DOS_Canonicalize(name.c_str(), buffer)) {
+			return permissive;
```
Can "permissive" be named better? (Even in the function body, it doesn't permit or deny any code paths, and instead only controls the return values if we failed to canoicalize).

It seems like three return types are desired? fully processed, couldn't canonicalize, and failed/illegal. (maybe an enum return is better than the taking in the permissive bool?) -- just throwing out ideas if they can help.

FeralChild64 replied:
Changed variable name to `allow_non_existing_files`. Enum return requires a definition + more complicated result checking.

MasterO2 commented:
Whenever `setver` is eventually merged into Dosbox Staging, the game below will be a good test for this newly supported command: 

https://github.com/dosbox-staging/dosbox-staging/wiki/Game-issues#jetfighter-the-adventure

I assume `setvar jf.exe 3.10` would be the syntax.  (`jf.exe` is JetFighter's executable).

FeralChild64 commented:
@kcgen True, we definitely need some protections against accidental wiping of the whole database.

@MasterO2 I would rather suggest something like `setver c:\games\jf\jf.exe 3.10` - this way we won't report old DOS version to some other software with the same executable name.

BTW, MS-DOS didn't support the path in `setver` command, but DR-DOS (which I was using as a teenager) did.

FeralChild64 commented:
@kcgen I have changed the code to require `/all` switch. Also added information about `/delete` and `/quiet` being same as `/d` and `/q` to the command help (longer switch versions are for DR-DOS compatibility only).

kcgen commented:
PVS is flagging a couple new hits in `setver`:
```
src/dos/program_setver.cpp : 2
```

zip is at the bottom here: https://github.com/dosbox-staging/dosbox-staging/actions/runs/4929529750?pr=2464

FeralChild64 commented:
@kcgen Should be fixed now.

kcgen commented:
Can't see anything left; passing tests, CI, and sanitizers. 
Thanks @FeralChild64 !
